paddle/fluid/imperative/reducer.h

```diff
@@ -0,0 +1,215 @@
+  } else {
+    operators::math::SplitFunctor<platform::CUDADeviceContext, T>
+        split_functor_;
+    split_functor_(context, *in, shape_refer, 0, &outs);
+  }
+}
+
+struct Group {
+  // Here, we use dense_contents & sparse_contents to
+  // achieve the tensor fuse. When is_sparse_ is true, sparse_contents work,
+  // conversely, dense_contents works. It is mutex relationship.
+  framework::Variable dense_contents;
+  framework::Variable* sparse_contents = nullptr;
+  bool is_sparse_ = false;
+
+  // for concat kernel
+  std::vector<framework::Tensor> dense_tensors;
+
+  std::vector<size_t> offset_;
+  std::vector<size_t> length_;
+  // Global indices of participating variables in the group
+  std::vector<size_t> variable_indices_;
+
+  // Number of params that haven't been ready. When it is 0, it means
+  // the group is ready.
+  size_t pending = -1;
+
+  // external message of group
+  framework::proto::VarType::Type dtype;
+
+  // context is used to select the stream for concat
+  void ConcatTensors(const platform::CUDADeviceContext& context) {
+    switch (dtype) {
+      case framework::proto::VarType::FP16:
+        ConcatTensorsForAllReduce<platform::float16>(context, dense_tensors,
+                                                     &dense_contents);
+        break;
+      case framework::proto::VarType::FP32:
+        ConcatTensorsForAllReduce<float>(context, dense_tensors,
+                                         &dense_contents);
+        break;
+      case framework::proto::VarType::FP64:
+        ConcatTensorsForAllReduce<double>(context, dense_tensors,
+                                          &dense_contents);
+        break;
+      default:
+        PADDLE_THROW(platform::errors::Unimplemented(
+            "Data type (%s) is not supported when it concats tensors for "
+            "allreduce.",
+            framework::DataTypeToString(dtype)));
+    }
+  }
+
+  // context is used to select the stream for split
+  void SplitTensors(const platform::CUDADeviceContext& context) {
+    switch (dtype) {
+      case framework::proto::VarType::FP16:
+        SplitTensorsForAllReduce<platform::float16>(context, &dense_contents,
+                                                    &dense_tensors);
+        break;
+      case framework::proto::VarType::FP32:
+        SplitTensorsForAllReduce<float>(context, &dense_contents,
+                                        &dense_tensors);
+        break;
+      case framework::proto::VarType::FP64:
+        SplitTensorsForAllReduce<double>(context, &dense_contents,
+                                         &dense_tensors);
+        break;
+      default:
+        PADDLE_THROW(platform::errors::Unimplemented(
+            "Data type (%s) is not supported when it splits tensors for "
+            "allreduce.",
+            framework::DataTypeToString(dtype)));
+    }
+  }
+};
+
+struct VariableIndex {
+  // record the index in groups_
+  size_t group_index;
+  size_t variable_index;
```

Comment:
`variable_index` may be a little bit confused, what about `inside_group_index`?

Reply:
done.